### PR TITLE
Compute q_num_blocks from kv_num_blocks if q_num_blocks is not passed in

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #130871
* #130869
* #130819
* #130818
* __->__ #130809

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/nn/attention/flex_attention.py [hunk 1]
```diff
@@ -121,6 +121,51 @@ def _no_mask(
...
+    out = create_dense_batched(num_blocks_in_row, col_indices)
+    return out
+
+
+def _dense_to_ordered(dense_mask) -> Tuple:
+    dense_mask = dense_mask.to(dtype=torch.int32)
+    num_blocks_in_row = dense_mask.sum(dim=-1)
+    col_indices = torch.argsort(dense_mask, dim=-1, descending=True, stable=True)
+    return (
+        num_blocks_in_row.to(torch.int32).to(dense_mask.device).contiguous(),
```
Comment (Skylion007) [inline torch/nn/attention/flex_attention.py:159]: nit: can / should these to `.to()` calls be fused into one? Or is that slower for some reason or can inductor not pick up on that?

File: test/inductor/test_flex_attention.py [hunk 2]
```diff
@@ -1209,46 +1269,73 @@ def test_captured_score_mod_aot_eager_gradcheck(
...
+            for ref, flex, golden in [
+                (out_ref, out_flex, out_golden),
+                (inputs[0].grad, inputs_flex[0].grad, inputs_golden[0].grad),
+                (inputs[1].grad, inputs_flex[1].grad, inputs_golden[1].grad),
+                (inputs[2].grad, inputs_flex[2].grad, inputs_golden[2].grad),
+            ]:
+                ref_error = rmse(ref, golden)
+                flex_error = rmse(flex, golden)
+                # Note: This has been carefully tested that FlexAttention is within
+                # 20% of the average error of SDPA! Do not bump this tolerance
```
Comment (drisspg) [inline test/inductor/test_flex_attention.py:1332]: lol

Comment (Skylion007) [reply test/inductor/test_flex_attention.py:1332]: What if we make torch SPDA even faster though... :P

File: torch/nn/attention/flex_attention.py [hunk 3]
```diff
@@ -167,25 +213,25 @@ class BlockMask:
     1. (kv_num_blocks, kv_indices): Used for the forwards pass of attention, as
     we reduce along the KV dimension.
 
-    2. (q_num_blocks, q_indices): Required for the backwards pass, as computing
-    dKV requires iterating along the mask along the Q dimension.
-
-    3. [OPTIONAL] (full_kv_num_blocks, full_kv_indices): This is optional and
+    2. [OPTIONAL] (full_kv_num_blocks, full_kv_indices): This is optional and
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:216]: Super nit not related to this PR, the "full" naming is somewhat confusing. When I think of other sparse formats the blocks are typically the non zero values. So in my mind when I think of the blocks I think full blocks are the blocks where every value is needed for computation, not that they are "full of zeros"

File: torch/nn/attention/flex_attention.py [hunk 4]
```diff
@@ -447,33 +486,20 @@ def _create_sparse_block_from_block_mask(
...
+        partial_bm = (None, None)
 
     return BlockMask(  # type: ignore[call-arg]
-        *full_bm,
-        *partial_bm,
-        KV_BLOCK_SIZE,
-        Q_BLOCK_SIZE,
-        mask_fn,
+        full_bm[0],
+        full_bm[1],
```
Comment (drisspg) [inline torch/nn/attention/flex_attention.py:497]: why not splat?

**Review by drisspg — APPROVED** [review]

File: torch/nn/attention/flex_attention.py (hunk 4 above)
Comment (Chillee) [reply torch/nn/attention/flex_attention.py:497]: It's smaller now and causes a bunch of mypy issues lol.

File: torch/nn/attention/flex_attention.py (hunk 1 above)
Comment (Chillee) [reply torch/nn/attention/flex_attention.py]: just removed the second one lol

**Pull request closed without merging**